File: app/validators.py
from __future__ import annotations

import re


def is_valid_package_name(value: str) -> bool:
    return bool(re.fullmatch(r'[a-zA-Z][a-zA-Z0-9_]*(\.[a-zA-Z][a-zA-Z0-9_]*)+', value))


def is_valid_version_code(value: str) -> bool:
    return bool(re.fullmatch(r'\d+', value))


def to_safe_file_stem(value: str) -> str:
    normalized = re.sub(r'[\\/:*?"<>|]', '-', value.strip())
    normalized = re.sub(r'\s+', ' ', normalized)[:80]
    return normalized or 'modded'

File: src/validators.ts
export function isValidPackageName(value: string): boolean {
  return /^[a-zA-Z][a-zA-Z0-9_]*(\.[a-zA-Z][a-zA-Z0-9_]*)+$/.test(value);
}

export function isValidVersionCode(value: string): boolean {
  return /^\d+$/.test(value);
}

export function toSafeFileStem(value: string): string {
  const normalized = value.trim().replace(/[\\/:*?"<>|]/g, '-').replace(/\s+/g, ' ').slice(0, 80);
  return normalized || 'modded';
}

export function normalizeRelPath(rawPath: string): string {
  const cleaned = (rawPath || '').replace(/\\/g, '/').trim();
  if (!cleaned || cleaned.startsWith('/')) {
    throw new Error('Invalid path');
  }
  const parts = cleaned.split('/').filter(part => part && part !== '.');
  if (parts.some(part => part === '..')) {
    throw new Error('Invalid path');
  }
  return parts.join('/');
}
